### ISSUE TITLE
Error in word_pos for WordAdapter

### ISSUE THREAD

**Issue by colobrodo** [vigna/dsi-bitstream-rs#18]
I have noted that when using a `BitReader` in conjunction with `WordAdapter` the `bit_pos()` positions are wrong as demostrated in this snippet:
 
```rust
use std::{
    fs::File,
    io::{BufReader, BufWriter},
};

use dsi_bitstream::prelude::*;

fn create_file(path: &str) -> anyhow::Result<()> {
    // create a bit writer on the file
    let mut writer = <BufBitWriter<NE, _>>::new(<WordAdapter<u64, _>>::new(
        BufWriter::with_capacity(1 << 20, File::create(path)?),
    ));

    println!("Write Gammas...");
    for n in 0..100 {
        writer.write_gamma(n)?;
    }
    writer.flush()?;
    Ok(())
}

fn main() -> anyhow::Result<()> {
    let path = "./mamma.bin";
    create_file(path)?;

    let reader = BufBitReader::<NE, _>::new(WordAdapter::<u32, _>::new(BufReader::new(
        File::open(path)?,
    )));
    let mut reader = CountBitReader::<_, _, false>::new(reader);
    for _ in 0..100 {
        let _ = reader.read_gamma()?;
        assert_eq!(
            reader.bits_read as u64,
            reader.bit_pos()?,
            "Number of bits read and position in bit stream are different"
        );
    }
    Ok(())
}

```

Looking at the implementation of `word_pos` in [`word_adapter.rs`](https://github.com/vigna/dsi-bitstream-rs/blob/7eeb8a33f7a82f1097baeb6c517a88ac652514d7/src/impls/word_adapter.rs#L95) and `bit_pos` in [`buf_bit_reader.rs`](https://github.com/vigna/dsi-bitstream-rs/blob/7eeb8a33f7a82f1097baeb6c517a88ac652514d7/src/impls/buf_bit_reader.rs#L330), I noticed that the former returns the position of the stream in *bytes* (using the [`stream_position`](https://doc.rust-lang.org/beta/std/io/trait.Seek.html#method.stream_position) function), while the latter requires a position relative to the size of the *word* to calculate the bit position.
To confirm this, the test above passes if `u8` is passed as a word (number of words and bytes are the same).
I think this may impact the [`offset_deg_iter`](https://docs.rs/webgraph/latest/webgraph/graphs/bvgraph/sequential/struct.BvGraphSeq.html#method.offset_deg_iter) iterator in webgraph and consequently the creation of offsets.

**Comment by zommiommy** [comment]
good catch!

**Comment by vigna** [comment]
Yes, really a good catch. This has been lurking in the code for almost two years now.

**Comment by zommiommy** [comment]
Luckly, the offsets were not affected as we load them with Mmap so we don't use that word adapter.

**Comment by vigna** [comment]
It's also really weird, I think we discussed and worked out precisely this issue a loooong time ago.

**Comment by zommiommy** [comment]
I'm sure we fixed an error like this before. This particular one is from Jenuary 27 2024 https://github.com/vigna/dsi-bitstream-rs/commit/15d5ce7858f5fa66c48576af99be3db9e923e12e

**Comment by vigna** [comment]
Ok, `stream_position`, but followed by an `align_to`. Don't you have that?

**Comment by zommiommy** [comment]
That commit was the introduction of align_to to handle partial reads. Before we only handled files whose length was a multiple of the word size.

**Comment by vigna** [comment]
Oh fuck. `align_to` alignes, it does not divide. How is it possible we never met this bug?

**Comment by vigna** [comment]
I guess the fuzzer was not looking at that method...

**Comment by zommiommy** [comment]
Yeah, I replaced it with `div_ceil` in the latest commit.

We don't use this in all of webgraph-rs CLI, so we didn't catch it.
The fuzzer tests `MemWordReader`, `MemWordWriterVec`, `MemWordWriterSlice`, but we don't test `WordAdapter`. I'll have to add more tests for it.

**Comment by vigna** [comment]
I guess we started to look for the other billion things that we had to do and we never did a proper coverage test. That's bad.

**Comment by vigna** [comment]
Why `div_ceil` and not just a division?

**Comment by zommiommy** [comment]
Yeah, we have it in the CI, but we only have 57% coverage. After the thesis, I'll have to increase it.

> Why `div_ceil` and not just a division?

Because we explicitly say we round up in the documentation:
```rust
/// Due to the necessity of managing files whose length is not a multiple of the
/// word length, [`read_word`](WordAdapter::read_word) will return a partially
/// read word extended with zeros at the end of such files.
///
/// To provide a sensible value after such a read,
/// [`word_pos`](WordAdapter::word_pos) will always return the position
/// of the underlying [`Seek`] rounded up to the next multiple of `W::Bytes`.
/// This approach, however, requires that if you adapt a [`Seek`], its current position must be
/// a multiple of `W::Bytes`, or the results of [`word_pos`](WordAdapter::word_pos)
/// will be shifted by the rounding.
```

**Comment by vigna** [comment]
Ahhh ok. `git blame` points at me as the culprit. So, I wrote the documentation correctly, but I miswrote the code.
